### ISSUE TITLE
[Bug] Customer & Vendor joins in transaction_details

### ISSUE THREAD

**Issue by jmongerlyra** [fivetran/dbt_netsuite#109]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the issue

There is a bug in `netsuite2__transaction_details model` that causes `company_name` and `vendor_name` to be NULL on certain lines.

The issue is in the joins to customer and vendor master tables. Certain transaction types can contain both transaction_lines.entity_id (line level) and transactions.entity_id (header level) values. The current join preferences the former in all cases.

However, in the example of a vendor bill with a billable customer referenced at the line level, this logic causes the customer ID to be joined to the vendor master table, resulting in a NULL. Instead, the join should be contingent on transaction type.

We will finalize a patch and submit a PR.

### Relevant error log or model output

```shell
n/a
```

### Expected behavior

Customer should always be populated on invoices and vendors always populated on bills.

### dbt Project configurations

```
config-version: 2
name: 'netsuite'
version: '0.12.0'
require-dbt-version: [">=1.3.0", "<2.0.0"]

models:
  netsuite:
    +materialized: table
    +schema: netsuite
    netsuite:
      intermediate:
        +materialized: ephemeral
    netsuite2:
      intermediate:
        +materialized: ephemeral

vars:
  netsuite:
    ## Netsuite staging models
    netsuite_accounting_books: "{{ ref('stg_netsuite__accounting_books') }}"
    netsuite_accounting_periods: "{{ ref('stg_netsuite__accounting_periods') }}"
    netsuite_accounts: "{{ ref('stg_netsuite__accounts') }}"
    netsuite_classes: "{{ ref('stg_netsuite__classes') }}"
    netsuite_consolidated_exchange_rates: "{{ ref('stg_netsuite__consolidated_exchange_rates') }}"
    netsuite_currencies: "{{ ref('stg_netsuite__currencies') }}"
    netsuite_customers: "{{ ref('stg_netsuite__customers') }}"
    netsuite_departments: "{{ ref('stg_netsuite__departments') }}"
    netsuite_expense_accounts: "{{ ref('stg_netsuite__expense_accounts') }}"
    netsuite_income_accounts: "{{ ref('stg_netsuite__income_accounts') }}"
    netsuite_items: "{{ ref('stg_netsuite__items') }}"
    netsuite_locations: "{{ ref('stg_netsuite__locations') }}"
    netsuite_subsidiaries: "{{ ref('stg_netsuite__subsidiaries') }}"
    netsuite_transaction_lines: "{{ ref('stg_netsuite__transaction_lines') }}"
    netsuite_transactions: "{{ ref('stg_netsuite__transactions') }}"
    netsuite_vendor_types: "{{ ref('stg_netsuite__vendor_types') }}"
    netsuite_vendors: "{{ ref('stg_netsuite__vendors') }}"
    netsuite2_account_types: "{{ ref('stg_netsuite2__account_types') }}"
    netsuite2_accounting_book_subsidiaries: "{{ ref('stg_netsuite2__accounting_book_subsidiaries') }}"
    netsuite2_accounting_books: "{{ ref('stg_netsuite2__accounting_books') }}"
    netsuite2_accounting_period_fiscal_calendars: "{{ ref('stg_netsuite2__accounting_period_fiscal_cal') }}"
    netsuite2_accounting_periods: "{{ ref('stg_netsuite2__accounting_periods') }}"
    netsuite2_accounts: "{{ ref('stg_netsuite2__accounts') }}"
    netsuite2_classes: "{{ ref('stg_netsuite2__classes') }}"
    netsuite2_consolidated_exchange_rates: "{{ ref('stg_netsuite2__consolidated_exchange_rates') }}"
    netsuite2_currencies: "{{ ref('stg_netsuite2__currencies') }}"
    netsuite2_customers: "{{ ref('stg_netsuite2__customers') }}"
    netsuite2_departments: "{{ ref('stg_netsuite2__departments') }}"
    netsuite2_entities: "{{ ref('stg_netsuite2__entities') }}"
    netsuite2_entity_address: "{{ ref('stg_netsuite2__entity_address') }}"
    netsuite2_items: "{{ ref('stg_netsuite2__items') }}"
    netsuite2_jobs: "{{ ref('stg_netsuite2__jobs') }}"
    netsuite2_location_main_address: "{{ ref('stg_netsuite2__location_main_address') }}"
    netsuite2_locations: "{{ ref('stg_netsuite2__locations') }}"
    netsuite2_subsidiaries: "{{ ref('stg_netsuite2__subsidiaries') }}"
    netsuite2_transaction_accounting_lines: "{{ ref('stg_netsuite2__transaction_accounting_lines') }}"
    netsuite2_transaction_lines: "{{ ref('stg_netsuite2__transaction_lines') }}"
    netsuite2_transactions: "{{ ref('stg_netsuite2__transactions') }}"
    netsuite2_vendor_categories: "{{ ref('stg_netsuite2__vendor_categories') }}"
    netsuite2_vendors: "{{ ref('stg_netsuite2__vendors') }}"
    accounts_pass_through_columns: []
    classes_pass_through_columns: []
    departments_pass_through_columns: []
    transactions_pass_through_columns: []
    transaction_lines_pass_through_columns: []
    balance_sheet_transaction_detail_columns: []
    income_statement_transaction_detail_columns: []
```

### Package versions

```
packages:
  - package: fivetran/netsuite_source
    version: [">=0.9.0", "<0.10.0"]
```

### What database are you using dbt with?

snowflake

### dbt Version

1.7.3

### Are you willing to open a PR to help address this issue?

- [X] Yes.
- [ ] Yes, but I will need assistance and will schedule time during our [office hours](https://calendly.com/fivetran-solutions-team/fivetran-solutions-team-office-hours) for guidance
- [ ] No.

**Comment by fivetran-jamie** [comment]
hey @jmongerlyra thanks for opening this and adding the fix to your PR!

left one question for you in the PR [here](https://github.com/fivetran/dbt_netsuite/pull/107/files#r1517870026) -- we'll be reviewing your PR this sprint FYI 🤠

**Comment by fivetran-avinash** [comment]
Thanks for addressing this issue in the existing PR @jmongerlyra ! I should hopefully pick up work on validating that all is well on this PR in this sprint.

**Comment by jmongerlyra** [comment]
I removed the patch from the existing PR. Needs some optimization. We will submit a separate PR once a new patch is available.

**Comment by fivetran-avinash** [comment]
No worries @jmongerlyra ! Looking forward for that PR.  

[I completed a first review of your PR, and we have a review open for your code](https://github.com/fivetran/dbt_netsuite/pull/107#pullrequestreview-1937633117) with some requested changes. 

Please go ahead and take a look. We're getting closer to having this ready to fold this into a future release of the package!

**Comment by fivetran-avinash** [comment]
Hi @jmongerlyra , hope all is well, just wanted to circle back here.

1)[ I believe you were going to submit a new PR](https://github.com/fivetran/dbt_netsuite/pull/107#discussion_r1518580302) after  the initial run was too slow. How is the status of that PR at the moment?
2) Any chance you were able to figure out how to handle [multiple self joins within dbt](https://github.com/fivetran/dbt_netsuite/pull/107#discussion_r1525433814)? I think that is a blocker for this current solution as well.

**Comment by jmongerlyra** [comment]
@fivetran-avinash Thanks for the following. Here's the current status of our fork against Fivetran main. https://github.com/fivetran/dbt_netsuite/compare/main...jmongerlyra:dbt_netsuite:main

> 1)https://github.com/fivetran/dbt_netsuite/pull/107#discussion_r1518580302 after the initial run was too slow. How is the status of that PR at the moment?

This is addressed in our fork. The issue is vendor and customer data lives in various fields depending on the transaction type. See [here](https://github.com/jmongerlyra/dbt_netsuite/blob/42b695021c4c056c5c309e13578d10deb44a27bd/models/netsuite2/netsuite2__transaction_details.sql#L163) for customers.

> 2) Any chance you were able to figure out how to handle https://github.com/fivetran/dbt_netsuite/pull/107#discussion_r1525433814? I think that is a blocker for this current solution as well.

We are still using the recursive CTE since that works in Snowflake.

**Comment by jmongerlyra** [comment]
@fivetran-avinash I went ahead and removed the recursive CTEs. See PR here to address this bug. https://github.com/fivetran/dbt_netsuite/pull/131

**Comment by fivetran-avinash** [comment]
Hey @jmongerlyra, thanks for raising this new PR last week! I can see you put a lot of thought and care into creating this out.  

This is quite a hefty PR to validate in its current state. Would it be possible for you to break up this PR into two separate PRs so we can better review and validate the changes for each:

* First PR:  The customer/vendor join updates in transaction details that were mentioned in this issue. We can also include the income statement update since it's quite small. We can move it into an upcoming sprint and prioritize this for release.
* Second PR: All other changes, primarily on balance sheet, since there are some complexities we'd like to address and it might take a little longer for us to confirm that your updates are the right approach to take.

Particularly around the second PR, we'd love if you could open up a new issue detailing the balance sheet updates, particularly regarding
* the `account_number` fixes for Retained Earnings and CTA
* the `is_account_leftside` fixes

Right now, it's not quite clear in this issue why these changes are being introduced.  An issue with supporting details for why these changes need to be made will make it easier for us to accelerate development. Then this will help us validate that your solution is the right one for all customers to use in a faster and more timely fashion. 

If you can open up that issue and separate out the PRs, then I think we will be ready to roll out the first PR ASAP and then make sure the second PR is addressed in a very near sprint. 

Hope this approach makes sense! Let me know if you have any questions, and thanks for your patience here.

**Comment by fivetran-avinash** [comment]
Thanks so much @jmongerlyra for splitting everything out!

I have some comments in the other issue you created, but I believe the plan is to bring this issue into a coming sprint so we can close it out very shortly! We will be in touch on this issue when we begin working on validating this bug fix.

**Comment by fivetran-avinash** [comment]
Thanks again for all the work you've put into separating out all these changes @jmongerlyra . Greatly appreciate our partnership in this work. After conducting some internal validation that these changes make sense, I've merged your PR into a release branch and will do our usual maintenance to ensure this PR is release ready. Will be in touch once these changes are live!

**Comment by fivetran-avinash** [comment]
Hi @jmongerlyra just circling back here and making sure these changes worked well for you!

**Comment by jmongerlyra** [comment]
@fivetran-avinash yep, we merged v0.14.0 into our fork.
